File: backend/internal/auth/token_bridge.go
package auth

import (
	"context"
	"crypto/rsa"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"math/big"
	"net/http"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

const (
	hsSecretMinLength     = 32
	jwksCacheTTLDefault   = 24 * time.Hour
	jwksClientTimeout     = 10 * time.Second
	bridgeTokenTTL        = 5 * time.Minute
	bridgeTokenTTLMinutes = 5
	base256Multiplier     = 256
)

// TokenBridge provides bidirectional authentication between HS256 and RS256 tokens
// Handles both WorkOS RS256 tokens (user auth) and internal HS256 service tokens
type TokenBridge struct {
	hsSecret      []byte
	jwksURL       string
	jwksClient    *http.Client
	jwksMutex     sync.RWMutex
	jwksCache     map[string]*rsa.PublicKey
	jwksCacheTime time.Time
	jwksCacheTTL  time.Duration
	audience      string
	issuer        string
}

// Note: WorkOSClaims is already defined in authkit_adapter.go
// Adding a TokenType field would require updating that struct

// JWKSResponse represents the JWKS endpoint response
type JWKSResponse struct {
	Keys []JWK `json:"keys"`
}

// JWK represents a JSON Web Key
type JWK struct {
	Kid string   `json:"kid"`
	Kty string   `json:"kty"`
	Use string   `json:"use"`
	Alg string   `json:"alg"`
	N   string   `json:"n"`
	E   string   `json:"e"`
	X5C []string `json:"x5_c,omitempty"`
}

// NewTokenBridge creates a new token bridge
// hsSecret: Shared HS256 secret for service tokens (min 32 bytes)
// jwksURL: WorkOS JWKS URL for RS256 public keys
// audience: Expected audience claim (typically WorkOS client ID)
// issuer: Expected issuer claim (typically WorkOS API base URL)
func NewTokenBridge(hsSecret []byte, jwksURL, audience, issuer string) (*TokenBridge, error) {
	if len(hsSecret) < hsSecretMinLength {
		return nil, errors.New("HS256 secret must be at least 32 bytes")
	}

	if jwksURL == "" {
		return nil, errors.New("JWKS URL is required")
	}

	tb := &TokenBridge{
		hsSecret:     hsSecret,
		jwksURL:      jwksURL,
		jwksCache:    make(map[string]*rsa.PublicKey),
		jwksCacheTTL: jwksCacheTTLDefault,
		audience:     audience,
		issuer:       issuer,
		jwksClient: &http.Client{
			Timeout: jwksClientTimeout,
		},
	}

	// Fetch JWKS on initialization
	if err := tb.RefreshJWKS(context.Background()); err != nil {
		slog.Error("Warning: Failed to fetch JWKS during initialization", "error", err)
		// Don't fail initialization - JWKS will be fetched on first validation
	}

	return tb, nil
}

// ValidateToken tries RS256 first (user tokens), falls back to HS256 (service tokens)
func (tb *TokenBridge) ValidateToken(ctx context.Context, tokenString string) (*WorkOSClaims, error) {
	// Try RS256 first (WorkOS user tokens)
	claims, err := tb.validateRS256Token(ctx, tokenString)
	if err == nil {
		return claims, nil
	}

	slog.Error("RS256 validation failed , trying HS256...", "error", err)

	// Fall back to HS256 (internal service tokens)
	claims, err = tb.validateHS256Token(tokenString)
	if err != nil {
		return nil, fmt.Errorf("token validation failed for both RS256 and HS256: %w", err)
	}

	return claims, nil
}

// validateRS256Token validates a WorkOS RS256 token using JWKS
func (tb *TokenBridge) validateRS256Token(ctx context.Context, tokenString string) (*WorkOSClaims, error) {
	tb.refreshJWKSIfStale(ctx)

	// Parse token without validation to extract kid
	token, _, err := new(jwt.Parser).ParseUnverified(tokenString, &WorkOSClaims{})
	if err != nil {
		return nil, fmt.Errorf("failed to parse token header: %w", err)
	}

	kid, ok := token.Header["kid"].(string)
	if !ok {
		return nil, errors.New("token missing kid header")
	}

	publicKey, err := tb.getPublicKey(ctx, kid)
	if err != nil {
		return nil, err
	}

	// Parse and validate token
	claims := &WorkOSClaims{}
	parsedToken, err := jwt.ParseWithClaims(tokenString, claims, func(token *jwt.Token) (interface{}, error) {
		// Verify signing method is RS256
		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return publicKey, nil
	})
	if err != nil {
		return nil, fmt.Errorf("failed to parse RS256 token: %w", err)
	}

	if !parsedToken.Valid {
		return nil, errors.New("invalid RS256 token")
	}

	if err := tb.validateClaims(claims); err != nil {
		return nil, err
	}

	return claims, nil
}

func (tb *TokenBridge) refreshJWKSIfStale(ctx context.Context) {
	tb.jwksMutex.RLock()
	cacheExpired := time.Since(tb.jwksCacheTime) > tb.jwksCacheTTL
	tb.jwksMutex.RUnlock()

	if cacheExpired {
		if err := tb.RefreshJWKS(ctx); err != nil {
			slog.Error("Warning: Failed to refresh JWKS", "error", err)
		}
	}
}

func (tb *TokenBridge) getPublicKey(ctx context.Context, kid string) (interface{}, error) {
	tb.jwksMutex.RLock()
	publicKey, exists := tb.jwksCache[kid]
	tb.jwksMutex.RUnlock()

	if exists {
		return publicKey, nil
	}

	if err := tb.RefreshJWKS(ctx); err != nil {
		return nil, fmt.Errorf("public key not found and JWKS refresh failed: %w", err)
	}

	tb.jwksMutex.RLock()
	publicKey, exists = tb.jwksCache[kid]
	tb.jwksMutex.RUnlock()

	if !exists {
		return nil, fmt.Errorf("public key not found for kid: %s", kid)
	}

	return publicKey, nil
}

func (tb *TokenBridge) validateClaims(claims *WorkOSClaims) error {
	if err := tb.validateAudience(claims); err != nil {
		return err
	}
	if err := tb.validateIssuer(claims); err != nil {
		return err
	}
	if err := tb.validateExpiry(claims); err != nil {
		return err
	}

	if claims.Subject == "" && claims.Sub != "" {
		claims.Subject = claims.Sub
	}

	return nil
}

func (tb *TokenBridge) validateAudience(claims *WorkOSClaims) error {
	// WorkOS AuthKit access tokens often omit aud; skip when claim missing.
	if tb.audience == "" || len(claims.Audience) == 0 {
		return nil
	}

	for _, aud := range claims.Audience {
		if aud == tb.audience {
			return nil
		}
	}

	return fmt.Errorf("invalid audience: expected %s, got %v", tb.audience, claims.Audience)
}

func (tb *TokenBridge) validateIssuer(claims *WorkOSClaims) error {
	if tb.issuer == "" || claims.Issuer == tb.issuer {
		return nil
	}

	return fmt.Errorf("invalid issuer: expected %s, got %s", tb.issuer, claims.Issuer)
}

func (tb *TokenBridge) validateExpiry(claims *WorkOSClaims) error {
	if claims.ExpiresAt != nil && claims.ExpiresAt.Before(time.Now()) {
		return fmt.Errorf("token expired at %v", claims.ExpiresAt.Time)
	}
	return nil
}

// validateHS256Token validates an internal HS256 service token
func (tb *TokenBridge) validateHS256Token(tokenString string) (*WorkOSClaims, error) {
	claims := &WorkOSClaims{}
	token, err := jwt.ParseWithClaims(tokenString, claims, func(token *jwt.Token) (interface{}, error) {
		// Verify signing method is HS256
		if token.Method.Alg() != "HS256" {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return tb.hsSecret, nil
	})
	if err != nil {
		return nil, fmt.Errorf("failed to parse HS256 token: %w", err)
	}

	if !token.Valid {
		return nil, errors.New("invalid HS256 token")
	}

	// Check expiry
	if claims.ExpiresAt != nil && claims.ExpiresAt.Before(time.Now()) {
		return nil, fmt.Errorf("token expired at %v", claims.ExpiresAt.Time)
	}

	// Service tokens should have type="service"
	if claims.TokenType != "service" {
		slog.Warn("Warning: HS256 token missing type=service claim")
	}

	// Sync RegisteredClaims.Subject from Sub field
	// (JWT parser populates Sub from "sub" claim, but RegisteredClaims.Subject is left empty)
	if claims.Subject == "" && claims.Sub != "" {
		claims.Subject = claims.Sub
	}

	return claims, nil
}

// CreateBridgeToken creates a short-lived HS256 service token (5 min TTL)
func (tb *TokenBridge) CreateBridgeToken(userID, orgID string) (string, error) {
	now := time.Now()
	claims := &WorkOSClaims{
		Sub:       userID,
		OrgID:     orgID,
		TokenType: "service",
		RegisteredClaims: jwt.RegisteredClaims{
			Subject:   userID,
			IssuedAt:  jwt.NewNumericDate(now),
			ExpiresAt: jwt.NewNumericDate(now.Add(bridgeTokenTTL)),
			NotBefore: jwt.NewNumericDate(now),
		},
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	tokenString, err := token.SignedString(tb.hsSecret)
	if err != nil {
		return "", fmt.Errorf("failed to sign bridge token: %w", err)
	}

	slog.Info("Created bridge token for user (org ), expires in minutes", "id", userID, "id", orgID, "id", bridgeTokenTTLMinutes)
	return tokenString, nil
}

// RefreshJWKS fetches latest public keys from WorkOS
func (tb *TokenBridge) RefreshJWKS(ctx context.Context) error {
	jwksResp, err := tb.fetchJWKSResponse(ctx)
	if err != nil {
		return err
	}

	newCache := buildRSAKeyCache(jwksResp.Keys)

	if len(newCache) == 0 {
		return errors.New("no valid RSA keys found in JWKS")
	}

	// Update cache
	tb.updateJWKSCache(newCache)

	return nil
}

func (tb *TokenBridge) fetchJWKSResponse(ctx context.Context) (JWKSResponse, error) {
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, tb.jwksURL, nil)
	if err != nil {
		return JWKSResponse{}, fmt.Errorf("failed to create JWKS request: %w", err)
	}

	resp, err := tb.jwksClient.Do(req)
	if err != nil {
		return JWKSResponse{}, fmt.Errorf("failed to fetch JWKS: %w", err)
	}
	defer func() {
		if err := resp.Body.Close(); err != nil {
			slog.Error("failed to close JWKS response body", "error", err)
		}
	}()

	if resp.StatusCode != http.StatusOK {
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return JWKSResponse{}, fmt.Errorf("failed to read JWKS response body: %w", err)
		}
		return JWKSResponse{}, fmt.Errorf(
			"JWKS endpoint returned status %d: %s",
			resp.StatusCode,
			string(body),
		)
	}

	var jwksResp JWKSResponse
	if err := json.NewDecoder(resp.Body).Decode(&jwksResp); err != nil {
		return JWKSResponse{}, fmt.Errorf("failed to decode JWKS response: %w", err)
	}

	return jwksResp, nil
}

func buildRSAKeyCache(keys []JWK) map[string]*rsa.PublicKey {
	newCache := make(map[string]*rsa.PublicKey)
	for _, jwk := range keys {
		if jwk.Kty != "RSA" {
			continue
		}

		publicKey, err := parseRSAPublicKey(jwk)
		if err != nil {
			slog.Error("Warning: Failed to parse JWK", "error", jwk.Kid, "error", err)
			continue
		}

		newCache[jwk.Kid] = publicKey
	}

	return newCache
}

func (tb *TokenBridge) updateJWKSCache(newCache map[string]*rsa.PublicKey) {
	tb.jwksMutex.Lock()
	tb.jwksCache = newCache
	tb.jwksCacheTime = time.Now()
	tb.jwksMutex.Unlock()
}

// parseRSAPublicKey converts a JWK to an RSA public key
func parseRSAPublicKey(jwk JWK) (*rsa.PublicKey, error) {
	// Decode modulus (n) - base64url encoded
	nBytes, err := base64.RawURLEncoding.DecodeString(jwk.N)
	if err != nil {
		return nil, fmt.Errorf("failed to decode modulus: %w", err)
	}

	// Decode exponent (e) - base64url encoded
	eBytes, err := base64.RawURLEncoding.DecodeString(jwk.E)
	if err != nil {
		return nil, fmt.Errorf("failed to decode exponent: %w", err)
	}

	// Convert modulus to big.Int
	modulus := new(big.Int).SetBytes(nBytes)

	// Convert exponent to int
	// The exponent is typically AQAB (65537) which fits in an int
	exponent := 0
	for _, byteValue := range eBytes {
		exponent = exponent*base256Multiplier + int(byteValue)
	}

	return &rsa.PublicKey{
		N: modulus,
		E: exponent,
	}, nil
}

// Close cleans up resources
func (tb *TokenBridge) Close() error {
	tb.jwksClient.CloseIdleConnections()
	return nil
}

File: src/tracertm/services/token_bridge.py
"""Token bridge for HS256/RS256 compatibility.

Provides bidirectional authentication between:
- WorkOS RS256 tokens (user authentication)
- Internal HS256 service tokens (service-to-service)
"""

from __future__ import annotations

import logging
import os
from datetime import UTC, datetime, timedelta
from typing import Any

import jwt
from jwt import PyJWKClient
from jwt.exceptions import ExpiredSignatureError, InvalidAudienceError, InvalidIssuerError

logger = logging.getLogger(__name__)

# Minimum secret length for HS256 tokens (256 bits = 32 bytes)
HS256_MIN_SECRET_LENGTH = 32


class TokenBridge:
    """Handles both WorkOS RS256 and internal HS256 token validation.

    Validation Strategy:
    1. Try RS256 first (WorkOS user tokens with JWKS)
    2. Fall back to HS256 (internal service tokens with shared secret)

    Token Types:
    - RS256: WorkOS AuthKit user authentication tokens
    - HS256: Short-lived service tokens (5 min TTL)
    """

    def __init__(
        self,
        hs_secret: str,
        jwks_url: str,
        audience: str | None = None,
        issuer: str | None = None,
        cache_ttl: int = 86400,  # 24 hours
    ) -> None:
        """Initialize token bridge.

        Args:
            hs_secret: Shared secret for HS256 service tokens (min 32 chars)
            jwks_url: WorkOS JWKS URL for RS256 public keys
            audience: Expected audience claim (typically WorkOS client ID)
            issuer: Expected issuer claim (typically WorkOS API base URL)
            cache_ttl: JWKS cache TTL in seconds (default 24 hours)
        """
        if len(hs_secret) < HS256_MIN_SECRET_LENGTH:
            msg = "HS256 secret must be at least 32 characters"
            raise ValueError(msg)

        self.hs_secret = hs_secret
        self.jwks_url = jwks_url
        self.audience = audience or os.getenv("WORKOS_CLIENT_ID")
        self.issuer = issuer
        self.jwks_client = PyJWKClient(
            jwks_url,
            cache_keys=True,
            lifespan=cache_ttl,
        )

    def validate_token(self, token: str) -> dict[str, Any]:
        """Validate token using RS256 (user) or HS256 (service).

        Args:
            token: JWT token string

        Returns:
            Decoded token claims

        Raises:
            jwt.InvalidTokenError: If token is invalid for both methods
        """
        # Read header (unverified) to avoid trying HS256 on an RS256 token
        try:
            unverified = jwt.get_unverified_header(token)
            alg = (unverified.get("alg") or "RS256").upper() if isinstance(unverified.get("alg"), str) else "RS256"
        except (jwt.InvalidTokenError, KeyError, ValueError):
            alg = "RS256"

        rs_failure: Exception | None = None
        if alg == "HS256":
            # Token claims HS256: try HS256 first, then RS256
            try:
                return self._validate_hs256_token(token)
            except (jwt.InvalidTokenError, ValueError) as e:
                rs_failure = e
                logger.debug("HS256 validation failed: %s, trying RS256...", e)
            try:
                return self._validate_rs256_token(token)
            except Exception as e:
                logger.exception("Token validation failed for both HS256 and RS256")
                msg = f"Token validation failed: HS256={rs_failure!r}, RS256={e!r}"
                raise jwt.InvalidTokenError(msg) from e

        # RS256 or other: try RS256 only (skip HS256 to avoid InvalidAlgorithmError)
        try:
            return self._validate_rs256_token(token)
        except ExpiredSignatureError as e:
            logger.warning("Token expired (RS256)")
            msg = "Token expired"
            raise jwt.InvalidTokenError(msg) from e
        except Exception as e:
            logger.debug("RS256 validation failed: %s", e)
            msg = f"Token validation failed: {e!r}"
            raise jwt.InvalidTokenError(msg) from e

    def _validate_rs256_token(self, token: str) -> dict[str, Any]:
        """Validate WorkOS RS256 token using JWKS.

        Uses optional audience/issuer verification so tokens without aud (e.g. some
        WorkOS AuthKit flows) still validate; when present, aud/iss are checked manually.
        """
        signing_key = self.jwks_client.get_signing_key_from_jwt(token)

        # Decode with signature and time checks only. Do not require aud/iss:
        # WorkOS AuthKit access tokens often omit aud; PyJWT defaults verify_aud to
        # verify_signature, so we pass audience=None/issuer=None and options explicitly.
        decoded = jwt.decode(
            token,
            signing_key.key,
            algorithms=["RS256"],
            audience=None,
            issuer=None,
            options={
                "verify_signature": True,
                "verify_exp": True,
                "verify_nbf": True,
                "verify_iat": True,
                "verify_aud": False,
                "verify_iss": False,
            },
        )

        # Optional manual issuer check when token has iss
        token_issuer = decoded.get("iss")
        if self.issuer and token_issuer:
            iss_ok = str(token_issuer).rstrip("/") == str(self.issuer).rstrip("/") or str(token_issuer).startswith(
                "https://api.workos.com/",
            )
            if not iss_ok:
                msg = f"Invalid issuer: expected {self.issuer!r}, got {token_issuer!r}"
                raise InvalidIssuerError(msg)

        # Optional manual audience check only when token has aud (avoid MissingRequiredClaimError)
        token_aud = decoded.get("aud")
        if self.audience and token_aud:
            if isinstance(token_aud, list[object]):
                if self.audience not in token_aud:
                    msg = f"Invalid audience: expected {self.audience!r}, got {token_aud!r}"
                    raise InvalidAudienceError(msg)
            elif token_aud != self.audience:
                msg = f"Invalid audience: expected {self.audience!r}, got {token_aud!r}"
                raise InvalidAudienceError(msg)

        logger.info("Validated RS256 token for user %s", decoded.get("sub"))
        return decoded

    def _validate_hs256_token(self, token: str) -> dict[str, Any]:
        """Validate internal HS256 service token.

        Args:
            token: JWT token string

        Returns:
            Decoded token claims

        Raises:
            jwt.InvalidTokenError: If token is invalid
        """
        decoded = jwt.decode(
            token,
            self.hs_secret,
            algorithms=["HS256"],
            options={
                "verify_signature": True,
                "verify_exp": True,
                "verify_nbf": True,
                "verify_iat": True,
            },
        )

        # Service tokens should have type="service"
        if decoded.get("type") != "service":
            logger.warning("HS256 token missing type=service claim")

        logger.info("Validated HS256 service token for user %s", decoded.get("sub"))
        return decoded

    def create_bridge_token(
        self,
        user_id: str,
        org_id: str,
        ttl_minutes: int = 5,
    ) -> str:
        """Create short-lived HS256 service token.

        Args:
            user_id: User ID (sub claim)
            org_id: Organization ID
            ttl_minutes: Token lifetime in minutes (default 5)

        Returns:
            Encoded JWT token string
        """
        now = datetime.now(UTC)
        payload = {
            "sub": user_id,
            "org_id": org_id,
            "type": "service",
            "iat": now,
            "exp": now + timedelta(minutes=ttl_minutes),
            "nbf": now,
        }

        token = jwt.encode(payload, self.hs_secret, algorithm="HS256")
        logger.info("Created bridge token for user %s (org %s), expires in %s minutes", user_id, org_id, ttl_minutes)
        return token

    def refresh_jwks(self) -> None:
        """Force refresh of JWKS cache.

        This is typically handled automatically by PyJWKClient,
        but can be called manually if needed.
        """
        # PyJWKClient handles caching internally
        # To force refresh, we'd need to recreate the client
        self.jwks_client = PyJWKClient(
            self.jwks_url,
            cache_keys=True,
            lifespan=86400,
        )
        logger.info("Refreshed JWKS cache")


def get_token_bridge() -> TokenBridge:
    """Factory function to create TokenBridge instance.

    Environment Variables:
        JWT_SECRET: Shared HS256 secret
        WORKOS_JWKS_URL: WorkOS JWKS endpoint
        WORKOS_CLIENT_ID: WorkOS client ID (audience)
        WORKOS_JWT_ISSUER: Expected issuer claim (optional)

    Returns:
        Configured TokenBridge instance

    Raises:
        ValueError: If required environment variables are missing
    """
    hs_secret = os.getenv("JWT_SECRET")
    if not hs_secret:
        msg = "JWT_SECRET environment variable is required"
        raise ValueError(msg)

    jwks_url = os.getenv("WORKOS_JWKS_URL")
    if not jwks_url:
        # Construct default JWKS URL from client ID
        client_id = os.getenv("WORKOS_CLIENT_ID")
        api_base = os.getenv("WORKOS_API_BASE_URL", "https://api.workos.com")
        if client_id:
            jwks_url = f"{api_base.rstrip('/')}/sso/jwks/{client_id}"
        else:
            msg = "Either WORKOS_JWKS_URL or WORKOS_CLIENT_ID must be set"
            raise ValueError(msg)

    return TokenBridge(
        hs_secret=hs_secret,
        jwks_url=jwks_url,
        audience=os.getenv("WORKOS_CLIENT_ID"),
        issuer=os.getenv("WORKOS_JWT_ISSUER"),
    )
